### ISSUE TITLE
critical_finalization fails on Mono Browser wasm Release @ Ubuntu.1804.Amd64.Open

### ISSUE THREAD

**Issue by EgorBo** [dotnet/runtime#75756]
https://dev.azure.com/dnceng-public/public/_build/results?buildId=19800&view=ms.vss-test-web.build-test-results-tab&runId=401084&paneView=debug

**Comment by EgorBo** [comment]
ah, it's probably a non-precise gc issue btw?

**Comment by vargaz** [comment]
It could be, on mono, its very hard to guarantee that a finalizer runs.

**Comment by kg** [comment]
Yeah I never managed to write a reliable test for it on the wasm binding side so I gave up.

**Comment by AntonLapounov** [comment]
> on mono, its very hard to guarantee that a finalizer runs.

Note that `P.count == 0` and `Q.count == 0` means none of 200 finalizers was executed.

**Comment by kg** [comment]
If I understand this situation right: It's wasm so there's no finalizer thread[1] so you would need to do a complex dance on the main thread at some point before exit where you do GCs repeatedly to fill the finalizer queue, then run all the finalizers on the queue. I'm sure we make a best effort to do something like this already but I'm not certain it would be possible to make it reliable. Wasm also scans the stack conservatively so any stray garbage on the stack could be keeping objects alive and preventing finalization.

[1] since threads are off by default right now

**Comment by lambdageek** [comment]
Could `GC.WaitForPendingFinalizers` just pump the finalizer queue synchronously on wasm? or is that just inviting deadlock?

As written, I agree this can't guarantee finalization on wasm

https://github.com/dotnet/runtime/blob/c7eb44d11bb5acca6ccc40de42314663cbaff145/src/tests/baseservices/critical_finalization/critical_finalization.cs#L38-L51

**Comment by lambdageek** [comment]
@lewing wishlist or wont fix, I think.

`GC.WaitForPendingFinalizers` cannot do the right thing on single-threaded wasm since there is no finalizer thread. I suggested pumping the queue synchronously, that's maybe worth trying, but I'm not sure it will do the right thing in general
